### PR TITLE
Minifier Prototype

### PR DESCRIPTION
**Pull request by Boshen**
To all reviewers, current and future collaborators: I want to make 100% sure that we all understand what I'm trying to achieve here.

I believe the current strengths we have right now are:

* A clear and easy to understand AST - see [ast/js.rs(https://github.com/Boshen/oxc/blob/main/crates/oxc_ast/src/ast/js.rs](https://github.com/Boshen/oxc/blob/main/crates/oxc_ast/src/ast/js.rs)
* A really fast parser - see [example benchmark](https://github.com/Boshen/oxc/actions/runs/4523376925#summary-12281305302)
* A [semantic analyzer](https://github.com/Boshen/oxc/blob/main/crates/oxc_semantic/src/builder.rs) for building symbols and scopes with a single AST pass
* Clean code - See implementation in this PR
* Focus on performance - we declare performance issues as bugs

Given the foundations mentioned above, we will be able to build a minifier that can outperform the [current benchmarks](https://github.com/privatenumber/minification-benchmarks).

* To build a minifier, the goal is super clear (compared to other toolings):
  * minimize execution time
  * minimize output text size 
* The minifier has larger problem domain but smaller individual problems. Collaborators can pick and focus on different components given their time availability and interest.
* This is much more fun to work on than the linter
* This is a standalone application with no outside interference
* No Plugins 😉 

----

A JavaScript minifier has three components:

1. printer
2. mangler
3. compressor

## Printer

The printer implementation resides in `oxc_printer`. It is responsible for removing whitespace.

## Mangler

The mangler implementation is part of the `SymbolTable` residing in `oxc_semantic`.
It is responsible for shortening variables. Its algorithm should be gzip friendly.

The printer is also responsible for printing out the shortened variable names.

## Compressor

The compressor is responsible for rewriting statements and expressions for minimal text output.
[Terser](https://github.com/terser/terser) is a good place to start for learning the fundamentals.

This is the prototype for the compressor. Printer is already in place, mangler will arrive shortly.

### CODE REVIEW

**Comment by Boshen** [comment]
To all collaborators and reviewers: I want to make 100% sure that we all understand what's going on here. 

* The minifier has larger problem domain but smaller individual problems. You can pick and focus on different things given your time availability and interest.
* The goal is super clear: minimize both execution time and minified output 
* This is much more fun to work on than the linter (I'll do the more boring tasks, e.g. porting more lint rules)
* No Plugins 😉 
* This is a standalone application which can gain larger adoption than the linter

**Review by shannonrothe — APPROVED** [review]
Great work, so exciting 🚀

File: crates/oxc_minifier/src/compressor/mod.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            let body = std::mem::replace(&mut while_stmt.body, dummy_body);
+            *stmt = self.ast.for_statement(SPAN, None, Some(test), None, body);
+        }
+    }
+
+    /* Expressions */
+
+    /// Transforms `undefined` => `void 0`
+    fn compress_undefined<'b>(&mut self, expr: &'b mut Expression<'a>) -> bool {
+        if expr.is_undefined() {
```
Comment (RiESAEX) [inline crates/oxc_minifier/src/compressor/mod.rs:137]: I think you need to check if it is shadowed.

**Comment by magic-akari** [comment]
> * To build a minifier, the goal is super clear (compared to other toolings):
>   
>   * minimize execution time
>   * minimize output text size

When they come into conflict, what actions do you take?
I have heard that cargo has an option called `opt-level`, which allows for precise control of optimization direction.

[ref](https://doc.rust-lang.org/cargo/reference/profiles.html#opt-level)

> - 0: no optimizations
> - 1: basic optimizations
> - 2: some optimizations
> - 3: all optimizations
> - "s": optimize for binary size
> - "z": optimize for binary size, but also turn off loop vectorization.

File: crates/oxc_minifier/src/compressor/mod.rs (hunk 1 above)
Comment (Boshen) [reply crates/oxc_minifier/src/compressor/mod.rs:137]: No worries, this is just a prototype.

**Comment by Boshen** [comment]
> I have heard that cargo has an option called opt-level, which allows for precise control of optimization direction.

Good suggestion, will keep an eye on this.

**Pull request merged**